proposals/init.md

```diff
@@ -0,0 +1,490 @@
+Init Only Members
+=====
+
+## Summary
+This proposal adds the concept of init-only members to C#. Members which can be
+written at the point of objection creation but become `readonly` once object
+creation has completed. This allows for a much more flexible immutable model
+in C#. 
+
+## Motivation
+The underlying mechanisms for building immutable data in C# haven't changed
+since 1.0. They remain:
+
+1. Declaring fields as `readonly`.
+1. Declaring properties that contain only a `get` accessor.
+
+These mechanisms are effective at allowing the construction of immutable data
+but they do so by adding cost to the boiler plate code of types and opting
+such types out of features like object and collection initializers. This means
+developers must choose between easy of use and immutability.
+
+Simple immutable object like `Point` requires twice as much boiler plate code
+to support construction as it does to declare the type. The bigger the type 
+the bigger the cost of this boiler plate:
+
+```cs
+struct Point
+{
+    public int X { get; }
+    public int Y { get; }
+
+    public Point(int X, int Y)
+    {
+        this.X = x;
+        this.Y = y;
+    }
+}
+```
+
+The `init` modifier makes immutable objects more flexible by allowing the
+caller to mutate the members during the act of construction. That means the 
+object can participate in object initialzers and thus removes the need for 
+all boiler plate code in the type. The `Point` type is now simply:
+
+```cs
+struct Point
+{
+    public int X { get; init; }
+    public int Y { get; init; }
+}
+```
+
+The consumer can then use object initializers to create the object
+
+```cs
+var p = new Point() { X = 42, Y = 13 };
+```
+
+## Detailed Design
+
+### init members
+An init only property is declared by using the `init` accessor in place of the 
+`set` accessor:
+
+```cs
+class Student
+{
+    public string FirstName { get; init; }
+    public string LastName { get; init; }
+}
+```
+
+An instance property containing an `init` accessor is considered settable in
+the following circumstances:
+
+- During an object initializer
+- Inside an instance constructor of the containing or derived type
```

Comment:
Don't want to get into validators here since they don't have a concrete design. Think it will be more appropriate for validators to call out the expanded places `init` accessors can be called just as the init only proposal calls out the expanded places `readonly` fields can be mutated.